core/src/main/scala/akka/persistence/Fallback.scala

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright (C) 2024 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+// Possible candidate for inclusion in Akka Persistence
+package akka.persistence
+
+import akka.actor.typed.ActorSystem
+import akka.actor.typed.Extension
+import akka.actor.typed.ExtensionId
+import akka.annotation.ApiMayChange
+import akka.util.Reflect
+
+import scala.annotation.nowarn
+import scala.concurrent.Future
+import scala.util.Failure
+import scala.util.Success
+import scala.util.control.NonFatal
+
+import java.time.Instant
+import java.util.concurrent.ConcurrentHashMap
+
+/**
+ * API MAY CHANGE
+ *
+ * Plugin API for a fallback store for events and snapshots. The journal and snapshot store implementations may, if
+ * configured to use an implementation of this API, use this fallback store to store events (resp. snapshots) while only
+ * storing a reference (aka "breadcrumb") returned by the plugin in the journal (resp. snapshot store). The event or
+ * snapshot can then be retrieved by providing the breadcrumb.
+ *
+ * Implementations must have a constructor taking an actor system, a Config object, and the config path
+ *
+ * @tparam Breadcrumb
+ *   the type of breadcrumb emitted and aceepted by this store: this type should have an Akka serializer registered and
+ *   may not be generic
+ */
+@ApiMayChange
+trait FallbackStore[Breadcrumb <: AnyRef] {
+  def toBreadcrumb(maybeBreadcrumb: AnyRef): Option[Breadcrumb]
+
+  def breadcumbClass: Class[Breadcrumb]
+
+  def loadEvent(
+      breadcrumb: Breadcrumb,
+      persistenceId: String,
+      seqNr: Long,
+      includePayload: Boolean): Future[Option[FallbackStore.EventFromFallback]]
+
+  def saveEvent(
+      persistenceId: String,
+      seqNr: Long,
+      serId: Int,
+      serManifest: String,
+      payload: Array[Byte]): Future[Breadcrumb]
+
+  def loadSnapshot(
+      breadcrumb: Breadcrumb,
+      persistenceId: String,
+      seqNr: Long): Future[Option[FallbackStore.SnapshotFromFallback]]
+
+  def saveSnapshot(
+      persistenceId: String,
+      seqNr: Long,
+      writeTimestamp: Instant,
+      eventTimestamp: Instant,
+      serId: Int,
+      serManifest: String,
+      payload: Array[Byte],
+      tags: Set[String],
+      meta: Option[((Int, String), Array[Byte])]): Future[Breadcrumb]
+}
+
+object FallbackStore {
+
+  /** API MAY CHANGE */
+  @ApiMayChange
+  final class EventFromFallback(val serId: Int, val serManifest: String, val payload: Option[Array[Byte]])
+
+  /** API MAY CHANGE */
+  @ApiMayChange
+  final class SnapshotFromFallback(
+      val seqNr: Long,
+      val writeTimestamp: Instant,
+      val eventTimestamp: Instant,
+      val serId: Int,
+      val serManifest: String,
+      val payload: Array[Byte],
+      val tags: Set[String],
+      val meta: Option[((Int, String), Array[Byte])])
```

Comment:
why are SnapshotFromFallback and EventFromFallback so different?

Reply:
For snapshots, because it's allowed for a store (regular or fallback) to only be able to store one snapshot ("update in place"), it's possible for the regular store to expect a snapshot from one sequence number and get a snapshot from a later sequence number.  An example of this would be:

* snapshot at seqNr A get saved to fallback
* breadcrumb gets saved for seqNr A
* snapshot at seqNr B later gets saved to fallback
* crash, breadcrumb not written
* breadcrumb for seqNr A read
* snapshot for seqNr B retrieved

Conversely, events are retrieved by persistence ID and sequence number, so there's no chance of getting an event other than the one you asked for, so the other data can stay in the journal.

Reply:
ok, I thought the s3 store was new coordinates for each seqNr, but I guess there is an advantage of in-place updates of s3 in the same way as for ordinary snapshots, which is that no garbage is ackumulated